lib/galaxy/tools/actions/__init__.py

```diff
@@ -88,8 +88,8 @@ def process_dataset(data, formats=None):
                 if not data.datatype.matches_any(formats):
                     # Need to refresh in case this conversion just took place, i.e. input above in tool performed the same conversion
                     trans.sa_session.refresh(data)
-                    target_ext, converted_dataset = data.find_conversion_destination(formats)
-                    if target_ext:
+                    direct_match, target_ext, converted_dataset = data.find_conversion_destination(formats)
```

Comment:
You don't need the `if not data.datatype.matches_any(formats)` statement above, you can replace that with 
```
direct_match, target_ext, converted_dataset = data.find_conversion_destination(formats)
if not direct_match:
````